src/conf_mode/nat64.py

```diff
@@ -0,0 +1,163 @@
+
+from operator import itemgetter
+import csv
+import json
+import os
+import re
+
+from vyos import ConfigError, airbag
+from vyos.config import Config
+from vyos.configdict import dict_merge
+from vyos.util import check_kmod, cmd, dict_search, run
+from vyos.xml import defaults
+
+airbag.enable()
+
+INSTANCE_REGEX = re.compile(r"instance-(\d+)")
+JOOL_CONFIG_DIR = "/run/jool"
+
+
+def get_config(config: Config | None = None) -> None:
+    """ """
+    if config is None:
+        conf = Config()
+
+    base = ["nat64"]
+    nat64 = conf.get_config_dict(base, key_mangling=("-", "_"), get_first_key=True)
+
+    # T2665: we must add the tagNode defaults individually until this is
+    # moved to the base class
+    for direction in ["source"]:
+        if direction in nat64:
+            default_values = defaults(base + [direction, "rule"])
+            if "rule" in nat64[direction]:
+                for rule in nat64[direction]["rule"]:
+                    nat64[direction]["rule"][rule] = dict_merge(
+                        default_values, nat64[direction]["rule"][rule]
+                    )
+
+    # Load in existing instances so we can destroy any unknown
+    lines = cmd("jool instance display --csv").splitlines()
+    for namespace, instance, mode in csv.reader(lines):
+        match = INSTANCE_REGEX.fullmatch(instance)
+        if not match:
+            # FIXME: Instances that don't match should be ignored but WARN'ed to the user
+            continue
+        num = match.group(1)
+
+        rules = nat64.setdefault("source", {}).setdefault("rule", {})
+        # Mark it for deletion
+        if num not in rules:
+            print(f"Deleting unknown instance: {num}")
+            rules[num] = {"deleted": True}
+            continue
+
+        # If the user changes the mode, recreate the instance else Jool fails with:
+        # Jool error: Sorry; you can't change an instance's framework for now.
+        if rules[num]["mode"] != mode:
+            print(f"Recreating instance {num} due to changed mode: {rules[num]['mode']} -> {mode}")
+            rules[num]["recreate"] = True
+
+    return nat64
+
+
+def verify(nat64) -> None:
+    """ """
+    if not nat64 or "deleted" in nat64:
+        # no need to verify the CLI as nat64 is going to be deactivated
+        return
+
+    # Verify that source.prefix is set and is a /96
+
+    return
+
+
+def generate(nat64) -> None:
+    """ """
+    os.makedirs(JOOL_CONFIG_DIR, exist_ok=True)
+
+    if dict_search("source.rule", nat64):
+        for rule, instance in dict_search("source.rule", nat64).items():
+            if "deleted" in instance:
+                continue
+
+            name = f"instance-{rule}"
+            config = {
+                "instance": name,
+                "framework": instance["mode"],
+                "global": {
+                    "pool6": instance["source"]["prefix"],
+                },
+                # "pool4": [],
+                # "bib": [],
+            }
```

Comment:
To generate config we usually use Jinja2 templates

Reply:
I'm in two minds about this. Jool's configuration is plain JSON, so either way, we need to produce valid json for it to consume.
I have a lot of experience with Saltstack and my extension Jinja. My experience is not a great. As a templating tool, Jinja is fine, but it's still just a templating tool at the end of the day. Templating is ultimately just a lazy way to construct computer-readable documents from text, but in all cases it's worse than constructing the data formally with an AST. An AST will always produce the correct result, whereas templating will often not. In the case of JSON which doesn't support niceties like trailing commas etc, I can't forsee any case where templating would be faster, less error prone or simpler than doing a direct conversion using the JSON library here.
I would urge that unless you have a very good reason to not do this, any alternative will be worse in almost every way.